File: src/python/GPMDS.py
#!/usr/bin/env python
from numpy import *
from MultiGPR import MultiGPR
import math
import time
import rospy

class GPMDS:
    def __init__(self, ell, sigmaF, sigmaN):
        self.ell = ell
        self.sigmaF = sigmaF
        self.sigmaN = sigmaN

        self.set_gp_parameters(ell, sigmaF, sigmaN)
        velocity_cap = rospy.get_param("velocity_cap")#0.1
        self.v_capLow = velocity_cap[0]
        self.v_capHigh = velocity_cap[1]

    def set_gp_parameters(self, ell, sigmaF, sigmaN):
        self.mGPR = MultiGPR(2,2)
        self.mGPR.setHyperParams(ell, sigmaF, sigmaN)

    def setOriginalDynamics(self, fun):
        # Set the original dynamics to the user passed function
        self.originalDynamics = fun

    def addData(self, position, velocity):
        theta = self.getTheta(position, velocity)
        # Add the data point to training data
        self.mGPR.addTrainingData(position, theta)

    def addDataM(self, position, velocity):
        for i in range(0,size(position[0])):
            self.addData(array([position[0,i],position[1,i]]), array([velocity[0,i],velocity[1,i]]))

        self.mGPR.prepareRegression()


    def getTheta(self, position, velocity):
        # Compute original dynamics at position
        originalVelocity = self.originalDynamics(position)
        # Find scaling between vectors
        kappa = (linalg.norm(velocity)/linalg.norm(originalVelocity))-1.0

        # Compare original velocity with velocity to compute speed scaling (kappa) and rotation axis angle
        # Check for divide by 0 error
        if linalg.norm(velocity)==0:
            normDesDir = velocity
        else:
            normDesDir = velocity/linalg.norm(velocity)

        #print originalVelocity
        if linalg.norm(originalVelocity)==0:
            normOrgDir = originalVelocity
        else:
            normOrgDir = originalVelocity/linalg.norm(originalVelocity)

        # Calculate angle of rotation
        angle = math.atan2(normDesDir[1],normDesDir[0])-math.atan2(normOrgDir[1],normOrgDir[0])

        # put in a good range
        if(angle > pi):
            angle = -(2*pi-angle)
        elif(angle < -pi):
            angle = 2*pi+angle;

        # Create the datapoint, theta (the 2 number output of the GPR- angle, scaling)
        theta = zeros(2)
        theta[0] = angle
        theta[1] = kappa

        return theta


    def removeData(self, start, end):#position, velocity):
        #print 'Removing data...'
        #theta = self.getTheta(position, velocity)
        s =  self.mGPR.removeTrainingData(start, end)#self.mGPR.removeTrainingData(position,theta)
        if len(self.mGPR.inputData) > 0:
            self.mGPR.prepareRegression()
        return s

    def reshapedDynamics(self, position):
        # Get original dynamics at position
        originalVelocity = self.originalDynamics(position)
        if self.mGPR.nData==0:
            velocity = originalVelocity
            #if(linalg.norm(velocity)>self.v_capHigh):
            #    velocity = velocity/linalg.norm(velocity)*self.v_capHigh
            #if (linalg.norm(velocity)<self.v_capLow):
            #    velocity = velocity/linalg.norm(velocity)*self.v_capLow
            #return velocity
        else:
            result = self.mGPR.doRegression(position)
            angle = result[0]
            kappa = result[1]
            kappa = max(kappa, -0.9)

            # Calculate the rotation matrix
            R = array([[cos(angle), -sin(angle)],[sin(angle),cos(angle)]])
            # Rotate velocity by R and scale by kappa
            velocity = (1.0 + kappa)*dot(R,originalVelocity)

        # Put velocity limits
        if (linalg.norm(velocity)>self.v_capHigh):
            velocity = velocity/linalg.norm(velocity)*self.v_capHigh
        if (linalg.norm(velocity)<self.v_capLow):
            velocity = velocity/linalg.norm(velocity)*self.v_capLow
        return velocity

File: GPMDS/src/GPMDS.cpp
#include "GPMDS/GPMDS.h"
#include <math.h>
#include "GPMDS/Timer.h"

using namespace Eigen;
using namespace std;

GPMDS::GPMDS(REALTYPE a)
{

}

//GPMDS::GPMDS(double a)
//{

//}

GPMDS::GPMDS(REALTYPE ell, REALTYPE sigmaF, REALTYPE sigmaN,REALTYPE speedErrorTol, REALTYPE angleErrorTol)
{
  //Initialize Gaussian Process
  mGPR = new MultiGPR(3,4);
  setGPParameters(ell, sigmaF, sigmaN);
  this->speedErrorTol = speedErrorTol;
  this->angleErrorTol = angleErrorTol;
  bSparse = false;
}

void GPMDS::setSparse(bool s=true){
  bSparse = s;
}

void GPMDS::setGPParameters(REALTYPE ell, REALTYPE sigmaF, REALTYPE sigmaN)
{
  //Create the GPR (input 3, output 4)
  //mGPR = new MultiGPR(3,4);
  //Set parameters for the GPR;
  mGPR->setHyperParams(ell, sigmaF, sigmaN);
}

void GPMDS::setOriginalDynamics(std::function<Vector3r(Vector3r)> fun)
{
  //Set the original dynamics to the function passed in by the user
  originalDynamics = fun;
}

void GPMDS::prepareFastQuery()
{
  //Set the original dynamics to the function passed in by the user
  mGPR->prepareRegression(true);
}

Vector4r GPMDS::computeLMDSParameters(Vector3r position, Vector3r velocity){
  Vector3r originalVelocity, axis;
  Vector4r theta;
  REALTYPE kappa, angle;

  //Compute original dynamics at position
  originalVelocity = originalDynamics(position);

  // Find the scaling between vectors
  kappa = (velocity.norm()/originalVelocity.norm())-1.0;
  //cout<<"velocoity.norm() "<<velocity.norm()<<" org. vel nomr. "<<originalVelocity.norm()<<endl;
  //Compare resulting original velocity with velocity to compute speed scaling (kappa) and rotation axis angle
  angle = acos((originalVelocity.dot(velocity))/(originalVelocity.norm()*velocity.norm()));
  Vector3r colinerity_measure;
  colinerity_measure= originalVelocity/originalVelocity.norm() - velocity/velocity.norm();
//  if(colinearity_measure.norm()<0.00001){
//      axis.setRandom();
//      axis -= axis.dot(velocity)*velocity/velocity.norm();
//  }else{
  axis = originalVelocity.cross(velocity);

  axis /= axis.norm();

  //Create the datapoint, theta (the 4 number output of the GPR- axis angle, scaling)
  for (int i=0; i<3; i++)
    theta(i) = angle*axis(i);

//  cout<<"start here"<<endl;
//  cout<<originalVelocity<<endl;
//  cout<<velocity<<endl;
//  cout<<axis<<endl;
//  cout<<"ang "<<angle<<endl;
  theta(3) = kappa;
//  for (int i = 0; i < 4; ++i) {
//      if(std::isnan(theta(i))){
//          cout<<"start here"<<endl;
//          cout<<originalVelocity<<endl;
//          cout<<velocity<<endl;
//          cout<<axis<<endl;
//          cout<<"ang "<<angle<<endl;
//      }
//  }

  return theta;
}


void GPMDS::addData(Vector3r position, Vector3r velocity)
{
  Vector4r theta;
  // need to subtract target posioitn
  theta = computeLMDSParameters(position, velocity);
  //cout<<"theta before GPR: "<<theta<<endl;
  if(checkNewData(position,theta))
    mGPR->addTrainingData(position, theta);
}

bool GPMDS::checkNewData(Vector3r position, Vector4r theta){
  if(mGPR->getNData() <1 || bSparse==false)
    return true;

  Vector4r theta_p;
  theta_p = mGPR->doRegression(position,true);
  Vector3r angleaxis_p,angleaxis;
  for(int i=0;i<3;i++){
    angleaxis(i) = theta(i);
    angleaxis_p(i) = theta_p(i);
  }
  REALTYPE angle,angle_p;
  angle_p = angleaxis_p.norm();
  angle = angleaxis.norm();
  if(fabs(theta_p(3) - theta(3))>speedErrorTol)// || fabs(angle_p-angle)>angleErrorTol){
  {
    return true;
  }
  else {
    cout << "Number of data points: " << mGPR->getNData() << endl;
    return false;
  }
}

void GPMDS::clearData()
{
    mGPR->clearTrainingData();
}

Vector3r GPMDS::reshapedDynamics(Vector3r position)
{

  Vector3r originalVelocity, velocity, axis;
  Vector4r result;
  Matrix3r rot_mat;
  REALTYPE kappa, angle;


  // Get original dynamics at position
  originalVelocity = originalDynamics(position);
  //cout<<"dsfijsijfd"<<endl;
  //cout<<originalVelocity<<endl;
  // Perform GPR on the given position to compute reshaping parameters
  //(*mGPR).prepareRegression();
  //cout<<originalVelocity<<endl;
  result = (*mGPR).doRegression(position);

  // Calculate axis and angle from reshaping parameters
  for (int i=0; i<3; i++){axis(i)=result(i);}
  kappa  = result(3);
  angle = axis.norm();
  axis = axis/angle;
  if (angle < 0.001)
  {
    // If the angle is very small, make rotation matrix the identity
   rot_mat.setIdentity();
   //cout<<"yes cam here"<<rot_mat<<endl;
  }
  else
  {
    // Find the rotation matrix from the axis and angle
    AngleAxisr rot_aa(angle, axis);
    rot_mat = rot_aa.toRotationMatrix();
  }
  //cout<<rot_mat<<endl;
  //Rotate and scale the original velocity
  velocity = rot_mat*originalVelocity;
  //cout<<velocity<<endl;
  //kappa = 0.0;
  if(kappa < -0.7)
      kappa = -0.7;

  kappa = 0.0;
  velocity= velocity*(kappa+1);
   // cout<<velocity<<endl;
  //Return resulting velocity
  return velocity;
}

Vector3r GPMDS::evaluateOriginalDynamics(Vector3r position)
{
    return originalDynamics(position);
}


void GPMDS::dummyFunction(int a){

}
